### PR TITLE
feat: Implement RegisteredServer#ping(PingOptions)

### LINKED ISSUE #937: RegisteredServer#ping does not support HEX colors
I was making a plugin to show to the admins of a server the motd of their servers, which uses the `RegisteredServer#ping` method, but when the ping is shown, it shows the colors as if they were shown to a Minecraft 1.15.2 or lower client.

This is because in the `#ping()` method the [ProtocolVersion.UNKNOWN is provided instead of any modern version](https://github.com/PaperMC/Velocity/blob/dev/3.0.0/proxy/src/main/java/com/velocitypowered/proxy/server/VelocityRegisteredServer.java#L88), which, when looking for the appropriate GsonSerializer for that version, the [Serializer for Legacy versions is used](https://github.com/PaperMC/Velocity/blob/f744b37ad5c7b6f00ca01681f26b75cad2584b7e/proxy/src/main/java/com/velocitypowered/proxy/VelocityServer.java#L750).

### PR DESCRIPTION
**Pull request by 4drian3d**
Resolves #937 by implementing a new method to ping a server with a specified options

### CODE REVIEW

**Comment by astei** [comment]
I'm not a huge fan of this - there's a lot of variables that can be tweaked for pinging a Minecraft server. I'd rather see a builder here, perhaps having an overload of `ping()` that takes in a `ServerPingOptions` that's constructed with a builder?

**Comment by 4drian3d** [comment]
> I'm not a huge fan of this - there's a lot of variables that can be tweaked for pinging a Minecraft server. I'd rather see a builder here, perhaps having an overload of `ping()` that takes in a `ServerPingOptions` that's constructed with a builder?

Personally I consider that `PingOptions` would be better because `ServerPingOptions` may generate confusion about the connection with `ServerPing` and its possible need to introduce this new class within the ServerPing class, although I am not sure of this name either.
The only additional field that I found useful to add was the timeout, otherwise, I don't know what other variable could be added.

On the topic of javadocs, I do not speak English natively, if someone can give a suggestion of any change, it would be perfect.

File: api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java
Comment (hugmanrique) [inline api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java]: "Contains the parameters used to ping a `RegisteredServer`". A `@see` tag to the `RegisteredServer#ping` method would be useful.

File: api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java
Comment (hugmanrique) [inline api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java]: "The protocol version used to ping the server", I would avoid mentioning emulation.

File: api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java
Comment (hugmanrique) [inline api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java]: "The maximum period of time to wait for a response from the remote server"

File: api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java
Comment (hugmanrique) [inline api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java]: Imo builder classes don't need getter methods

File: api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java
Comment (hugmanrique) [inline api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java]: We need to provide a default timeout value instead of throwing here. We can use 0 (see below).

File: api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java
Comment (hugmanrique) [inline api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java]: Check `time` is positive or `-1` for no timeout; durations can be negative. Perhaps we can store the duration as milliseconds directly in the builder as well to simplify the `-1` case (there's no concept of an unlimited `Duration`).

File: proxy/src/main/java/com/velocitypowered/proxy/server/VelocityRegisteredServer.java
Comment (hugmanrique) [inline proxy/src/main/java/com/velocitypowered/proxy/server/VelocityRegisteredServer.java]: To clean this a bit, we can make the no-arguments `ping` method pass a default `PingOptions` instance (which can be cached and be exposed as an static member called `DEFAULT` in that class) to this method, and remove the method with all options below (which won't "scale" as the option set grows in the future).

File: proxy/src/main/java/com/velocitypowered/proxy/server/VelocityRegisteredServer.java
Comment (hugmanrique) [inline proxy/src/main/java/com/velocitypowered/proxy/server/VelocityRegisteredServer.java]: -1 usually means "no timeout", could we use 0 for the default value in the `PingOptions` class?

File: api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java
Comment (hugmanrique) [inline api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java:56]: Some users prefer the convention `PingOptions.builder().[...].build().ping(server)` instead of `server.ping(PingOptions.builder().[...].build())`. Can we add that helper method that forwards the call to the existing ping member in `RegisteredServer`? I must say this is a big nit, however.

**Review by hugmanrique — COMMENTED** [review]
Made some suggestions regarding the javadocs and some methods. They are pretty minor, everything else LGTM.

File: api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java
Comment (4drian3d) [reply api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java]: Initially I added them because I was implementing the `PingOptions#asBuilder` method, _similar to `ServerPing#asBuilder`_. But I think in this class it won't be necessary for the moment.

File: api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java
Comment (4drian3d) [reply api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java:56]: I'm not so sure about adding that method, I don't see much need to add it, besides that in Velocity there is no class that is built by builder that behaves that way

Comment (hugmanrique) [reply api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java:57]: I agree :) Let's ignore it then.

File: proxy/src/main/java/com/velocitypowered/proxy/server/VelocityRegisteredServer.java
Comment (4drian3d) [reply proxy/src/main/java/com/velocitypowered/proxy/server/VelocityRegisteredServer.java]: I'm not sure how good it would be to leave the option of no timeout, although I'm not an expert on the subject to be able to give an opinion on it

File: api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java
Comment (hugmanrique) [inline api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java:112]: Check for null here and initialize the version class field to `UNKNOWN`; that way we don't need the null check in the `build` method.

File: api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java
Comment (hugmanrique) [inline api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java]: Let's not document the specific default timeout, as it may be changed by the user.

**Review by hugmanrique — COMMENTED** [review]
Some last changes, thank you.

File: api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java
Comment (hugmanrique) [reply api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java:56]: I agree, let's keep it this way :)

File: proxy/src/main/java/com/velocitypowered/proxy/server/VelocityRegisteredServer.java
Comment (hugmanrique) [reply proxy/src/main/java/com/velocitypowered/proxy/server/VelocityRegisteredServer.java]: In the end, some effective timeout is placed by the network equipment. I was noting that the value -1 is usually used to mean "no limit", whereas 0 sometimes means "use a sane default".

Comment (4drian3d) [reply proxy/src/main/java/com/velocitypowered/proxy/server/VelocityRegisteredServer.java]: Ok, I removed the limits for the timeout and documented the meaning of a value of 0 and a negative value

File: api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java
Comment (hugmanrique) [inline api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java]: This message needs to be updated, everything else LGTM! 😀

Comment (4drian3d) [reply api/src/main/java/com/velocitypowered/api/proxy/server/PingOptions.java]: Fixed 😅

**Comment by hugmanrique** [comment]
For future reference (and curiosity), Netty's `ReadTimeoutHandler` uses [`IdleStateHandler`](https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/timeout/IdleStateHandler.java) internally. While not documented, passing a negative value to the timeout handler thus disables the timeout logic (see the constructors of the latter class, which while their javadoc only specifies this is the case for zero, apply the same logic for negative values). This is why we can use a negative integer as the "no timeout" value in `PingOptions`, and handle the zero case separately.

**Review by hugmanrique — APPROVED** [review]

**Review by astei — APPROVED** [review]

**Pull request merged**
